### ISSUE TITLE
"Writing large partition reaper_db/node_metrics_v2" Warning

### ISSUE THREAD

**Issue by dave-pollock** [thelastpickle/cassandra-reaper#850]
Hi. I've recently deployed Reaper to a 6 node Cassandra cluster. Reaper is running in sidecar mode.

Shortly after Reaper was deployed we began to see warnings such as the following logged frequently in the Cassandra logs:
```
WARN  [PerDiskMemtableFlushWriter_0:57] 2020-02-14 00:57:46,510 BigTableWriter.java:211 - Writing large partition reaper_db/node_metrics_v2:prod-1:org.apache.cassandra.metrics:ClientRequest:2020021400 (130.063MiB) to sstable /data/cassandra/data/reaper_db/node_metrics_v2-5524a1304dfd11eaaacd1b3a8cc4fc6c/md-27-big-Data.db
```
I've tried to count the rows in the node_metrics_v2 table, but the query times out, even when timeouts are increased to 60 seconds, so I assume there is a lot of data in this table.

The cluster itself is very lightly used - it's not in production yet so there is only one keyspace besides the system keyspaces and reaper keyspace. Certainly not talking about massive scale here.

Is this expected behaviour?

We're using Reaper v2.0.2.

Our cassandra-reaper.yaml: https://gist.github.com/dave-pollock/a278ff6efcdedabff0e3c4ac3578eddd

<details>
<summary>nodetool version</summary>

```
ReleaseVersion: 3.11.5
```
</details>

<details>
<summary>nodetool cfstats -- reaper_db.node_metrics_v2</summary>

```
Total number of tables: 58
----------------
Keyspace : reaper_db
	Read Count: 4657
	Read Latency: 0.14297036718917758 ms
	Write Count: 55245
	Write Latency: 0.0352549733007512 ms
	Pending Flushes: 0
		Table: node_metrics_v2
		SSTable count: 6
		Space used (live): 411803116
		Space used (total): 411803116
		Space used by snapshots (total): 0
		Off heap memory used (total): 5604419
		SSTable Compression Ratio: 0.17955615395293403
		Number of partitions (estimate): 32
		Memtable cell count: 54648
		Memtable data size: 5810912
		Memtable off heap memory used: 5327014
		Memtable switch count: 0
		Local read count: 0
		Local read latency: NaN ms
		Local write count: 54648
		Local write latency: 0.040 ms
		Pending flushes: 0
		Percent repaired: 91.76
		Bloom filter false positives: 0
		Bloom filter false ratio: 0.00000
		Bloom filter space used: 136
		Bloom filter off heap memory used: 88
		Index summary off heap memory used: 485
		Compression metadata off heap memory used: 276832
		Compacted partition minimum bytes: 4055270
		Compacted partition maximum bytes: 155469300
		Compacted partition mean bytes: 58676075
		Average live cells per slice (last five minutes): NaN
		Maximum live cells per slice (last five minutes): 0
		Average tombstones per slice (last five minutes): NaN
		Maximum tombstones per slice (last five minutes): 0
		Dropped Mutations: 0

----------------
```
</details>

<details>
<summary>DESCRIBE KEYSPACE reaper_db;</summary>

```
CREATE KEYSPACE reaper_db WITH replication = {'class': 'NetworkTopologyStrategy', 'europe-west1': '3'}  AND durable_writes = false;

CREATE TABLE reaper_db.diagnostic_event_subscription (
    id uuid PRIMARY KEY,
    cluster text,
    description text,
    events set<text>,
    export_file_logger text,
    export_http_endpoint text,
    export_sse boolean,
    nodes set<text>
) WITH bloom_filter_fp_chance = 0.1
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.LeveledCompactionStrategy'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';

CREATE TABLE reaper_db.running_reapers (
    reaper_instance_id uuid PRIMARY KEY,
    last_heartbeat timestamp,
    reaper_instance_host text
) WITH bloom_filter_fp_chance = 0.1
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.LeveledCompactionStrategy'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 180
    AND gc_grace_seconds = 180
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';

CREATE TABLE reaper_db.repair_unit_v1 (
    id timeuuid PRIMARY KEY,
    blacklisted_tables set<text>,
    cluster_name text,
    column_families set<text>,
    datacenters set<text>,
    incremental_repair boolean,
    keyspace_name text,
    nodes set<text>,
    repair_thread_count int
) WITH bloom_filter_fp_chance = 0.1
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.LeveledCompactionStrategy'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';

CREATE TABLE reaper_db.node_operations (
    cluster text,
    type text,
    time_bucket text,
    host text,
    data text,
    ts timestamp,
    PRIMARY KEY ((cluster, type, time_bucket), host)
) WITH CLUSTERING ORDER BY (host ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.TimeWindowCompactionStrategy', 'compaction_window_size': '1', 'compaction_window_unit': 'HOURS', 'max_threshold': '32', 'min_threshold': '4', 'unchecked_tombstone_compaction': 'true'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 300
    AND gc_grace_seconds = 300
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';

CREATE TABLE reaper_db.repair_schedule_by_cluster_and_keyspace (
    cluster_name text,
    keyspace_name text,
    repair_schedule_id timeuuid,
    PRIMARY KEY ((cluster_name, keyspace_name), repair_schedule_id)
) WITH CLUSTERING ORDER BY (repair_schedule_id ASC)
    AND bloom_filter_fp_chance = 0.1
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.LeveledCompactionStrategy'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';

CREATE TABLE reaper_db.repair_run_by_cluster (
    cluster_name text,
    id timeuuid,
    PRIMARY KEY (cluster_name, id)
) WITH CLUSTERING ORDER BY (id ASC)
    AND bloom_filter_fp_chance = 0.1
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.LeveledCompactionStrategy'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';

CREATE TABLE reaper_db.repair_schedule_v1 (
    id timeuuid PRIMARY KEY,
    creation_time timestamp,
    days_between int,
    intensity double,
    next_activation timestamp,
    owner text,
    pause_time timestamp,
    repair_parallelism text,
    repair_unit_id timeuuid,
    run_history set<timeuuid>,
    segment_count int,
    segment_count_per_node int,
    state text
) WITH bloom_filter_fp_chance = 0.1
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.LeveledCompactionStrategy'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';

CREATE TABLE reaper_db.schema_migration (
    applied_successful boolean,
    version int,
    executed_at timestamp,
    script text,
    script_name text,
    PRIMARY KEY (applied_successful, version)
) WITH CLUSTERING ORDER BY (version ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy', 'max_threshold': '32', 'min_threshold': '4'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';

CREATE TABLE reaper_db.cluster (
    name text PRIMARY KEY,
    last_contact timestamp,
    partitioner text,
    properties text,
    seed_hosts set<text>,
    state text
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.LeveledCompactionStrategy'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';

CREATE TABLE reaper_db.schema_migration_leader (
    keyspace_name text PRIMARY KEY,
    leader uuid,
    leader_hostname text,
    took_lead_at timestamp
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy', 'max_threshold': '32', 'min_threshold': '4'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';

CREATE TABLE reaper_db.snapshot (
    cluster text,
    snapshot_name text,
    cause text,
    creation_time timestamp,
    owner text,
    PRIMARY KEY ((cluster, snapshot_name))
) WITH bloom_filter_fp_chance = 0.1
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.LeveledCompactionStrategy'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';

CREATE TABLE reaper_db.node_metrics_v1 (
    run_id uuid,
    time_partition bigint,
    node text,
    active_anticompactions int,
    cluster text,
    datacenter text,
    has_repair_running boolean,
    pending_compactions int,
    requested boolean,
    PRIMARY KEY ((run_id, time_partition), node)
) WITH CLUSTERING ORDER BY (node ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.TimeWindowCompactionStrategy', 'compaction_window_size': '2', 'compaction_window_unit': 'MINUTES', 'max_threshold': '32', 'min_threshold': '4', 'unchecked_tombstone_compaction': 'true'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 180
    AND gc_grace_seconds = 120
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';

CREATE TABLE reaper_db.node_metrics_v2 (
    cluster text,
    metric_domain text,
    metric_type text,
    time_bucket text,
    host text,
    metric_scope text,
    metric_name text,
    ts timestamp,
    metric_attribute text,
    value double,
    PRIMARY KEY ((cluster, metric_domain, metric_type, time_bucket), host, metric_scope, metric_name, ts, metric_attribute)
) WITH CLUSTERING ORDER BY (host ASC, metric_scope ASC, metric_name ASC, ts DESC, metric_attribute ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.TimeWindowCompactionStrategy', 'compaction_window_size': '1', 'compaction_window_unit': 'HOURS', 'max_threshold': '32', 'min_threshold': '4', 'unchecked_tombstone_compaction': 'true'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 864000
    AND gc_grace_seconds = 300
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';

CREATE TABLE reaper_db.repair_run (
    id timeuuid,
    segment_id timeuuid,
    cause text static,
    cluster_name text static,
    coordinator_host text,
    creation_time timestamp static,
    end_time timestamp static,
    end_token varint,
    fail_count int,
    intensity double static,
    last_event text static,
    owner text static,
    pause_time timestamp static,
    repair_parallelism text static,
    repair_unit_id timeuuid static,
    segment_count int static,
    segment_end_time timestamp,
    segment_start_time timestamp,
    segment_state int,
    start_time timestamp static,
    start_token varint,
    state text static,
    tables set<text> static,
    token_ranges text,
    PRIMARY KEY (id, segment_id)
) WITH CLUSTERING ORDER BY (segment_id ASC)
    AND bloom_filter_fp_chance = 0.1
    AND caching = {'keys': 'ALL', 'rows_per_partition': '5000'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.LeveledCompactionStrategy'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';

CREATE TABLE reaper_db.repair_run_by_unit (
    repair_unit_id timeuuid,
    id timeuuid,
    PRIMARY KEY (repair_unit_id, id)
) WITH CLUSTERING ORDER BY (id ASC)
    AND bloom_filter_fp_chance = 0.1
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.LeveledCompactionStrategy'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';

CREATE TABLE reaper_db.leader (
    leader_id timeuuid PRIMARY KEY,
    last_heartbeat timestamp,
    reaper_instance_host text,
    reaper_instance_id uuid
) WITH bloom_filter_fp_chance = 0.1
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.LeveledCompactionStrategy'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 600
    AND gc_grace_seconds = 600
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';
```
</details>

Any input appreciated. Thanks!

**Comment by adejanovski** [comment]
Hi @dave-pollock,

thanks for reporting this. It looks like we don't have the right partitioning here. I'll fix it ASAP and issue a new release as large partitions aren't something we want to generate in Cassandra.

**Comment by dave-pollock** [comment]
Good to know there's not something wrong with our setup! Thanks @adejanovski

**Comment by adejanovski** [comment]
Fixed in v2.0.3 that was just released
